File: game/scripts/vscripts/clases/pre_round.lua
PreRound = PreRound or class({}, nil, GameState)

local PRE_ROUND_TIMER = PRE_ROUND_DURATION

function PreRound:constructor(alliances)
    getbase(PreRound).constructor(self, alliances)
    self.time_remaining = PRE_ROUND_TIMER * 30
    self.radiant_spawn = Entities:FindByName(nil, "radiant_spawn")
    self.dire_spawn = Entities:FindByName(nil, "dire_spawn")

    
    for _,player in pairs(self:GetAllPlayers()) do
        local hero = player.hero
        hero:AddNewModifier(hero, nil, 'modifier_restricted', {})
    end
end


function PreRound:Update()
    if self.time_remaining > 0 then
        self.time_remaining = self.time_remaining - 1

        self:UpdateGameTimer(math.floor(self.time_remaining/30))
        if self.time_remaining <= 0 then
            self:EndPreRound()
        end
    end
end

function PreRound:EndPreRound()
    for _,player in pairs(self:GetAllPlayers()) do
        local target = nil

        if player.alliance.name == "DOTA_ALLIANCE_RADIANT" then
            target = self.radiant_spawn
        end
        if player.alliance.name == "DOTA_ALLIANCE_DIRE" then
            target = self.dire_spawn
        end

        local hero = player.hero

        FindClearSpaceForUnit(hero, target:GetAbsOrigin(), true)
        CustomEntities:Reset(hero)
        CustomEntities:AutoUpgradeAbilities(hero)
        hero:RemoveModifierByName('modifier_restricted')
        hero:AddNewModifier(hero, nil, "modifier_adrenaline", { duration = ADRENALINE_DURATION })
    end

    GameRules.GameMode.pre_round = nil
    GameRules.GameMode.round = Round(self.alliances)
    GameRules.GameMode:SetState(CustomGameState.ROUND_IN_PROGRESS)
end

File: src/vscripts/clases/pvp/pre_round.ts
import { CustomModifier } from "../../abilities/framework/custom_modifier";
import { Alliance } from "../../alliances";
import { registerModifier } from "../../lib/dota_ts_adapter";
import { ModifierAdrenaline } from "../../modifiers/modifier_adrenaline";
import GameState, { CustomGameState } from "../game_state";

@registerModifier("modifier_restricted")
class ModifierRestricted extends CustomModifier<undefined> {
    IsHidden() {
        return true;
    }

    IsDebuff() {
        return false;
    }

    IsPurgable() {
        return false;
    }

    RemoveOnDeath() {
        return false;
    }

    CheckState() {
        return {
            [ModifierState.COMMAND_RESTRICTED]: true
        };
    }
}

export default class PreRound extends GameState {
    radiant_spawn: Vector;
    dire_spawn: Vector;

    constructor(alliances: Alliance[], duration: number) {
        super(alliances, duration);
        this.radiant_spawn = Entities.FindByName(undefined, "radiant_spawn")!.GetAbsOrigin();
        this.dire_spawn = Entities.FindByName(undefined, "dire_spawn")!.GetAbsOrigin();

        this.GetAllPlayers().forEach(player => {
            const hero = player.entity?.handle;
            if (hero) {
                ModifierRestricted.apply(hero, hero, undefined, {});
            }
        });
    }

    Update(): void {
        super.Update();
        if (this.timeRemaining >= 0) {
            this.UpdateGameTimer(math.floor(this.timeRemaining / 30));
        }
        if (this.timeRemaining == 0) {
            this.EndPreRound();
        }
    }

    EndPreRound(): void {
        this.GetAllPlayers().forEach(player => {
            const hero = player.entity?.handle;
            if (hero) {
                if (player.alliance) {
                    const allianceName = player.alliance.id;
                    let target = undefined;

                    if (allianceName == AllianceId.radiant) {
                        target = this.radiant_spawn;
                    } else if (allianceName == AllianceId.dire) {
                        target = this.dire_spawn;
                    }

                    FindClearSpaceForUnit(hero, target!, true);
                    CustomEntitiesLegacy.Reset(hero);
                    CustomEntitiesLegacy.AutoUpgradeAbilities(hero);
                    hero.RemoveModifierByName("modifier_restricted");
                    ModifierAdrenaline.apply(hero, hero, undefined, {
                        // duration: ADRENALINE_DURATION
                    });
                }
            }
        });

        GameRules.Addon.SetState(CustomGameState.ROUND_IN_PROGRESS);
    }
}
